web/Services/PostcodeService.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System.Text.RegularExpressions;
+using BricksAndHearts.Database;
+using BricksAndHearts.Models;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace BricksAndHearts.Services;
+
+public interface IPostcodeService
+{
+    public string FormatPostcode(string postcode);
+    public Task AddSinglePostcodeToDatabaseIfAbsent(string postcode);
+    public Task BulkAddPostcodesToDatabaseIfAbsent(List<string> postcodes);
+}    
+
+
+public class PostcodeService : IPostcodeService
+{
+    private readonly HttpClient _client;
+    private readonly BricksAndHeartsDbContext _dbContext;
+    private readonly ILogger<PostcodeService> _logger;
+
+    public PostcodeService(ILogger<PostcodeService> logger, BricksAndHeartsDbContext dbContext, HttpClient client)
+    {
+        _logger = logger;
+        _dbContext = dbContext;
+        _client = client;
+    }
+
+    public string FormatPostcode(string postcode)
+    {
+        if (Regex.IsMatch(postcode,
+                @"([Gg][Ii][Rr] 0[Aa]{2})|((([A-Za-z][0-9]{1,2})|(([A-Za-z][A-Ha-hJ-Yj-y][0-9]{1,2})|(([A-Za-z][0-9][A-Za-z])|([A-Za-z][A-Ha-hJ-Yj-y][0-9][A-Za-z]?))))\s?[0-9][A-Za-z]{2})"))
+        {
+            postcode = Regex.Replace(postcode, @"^(\S+?)\s*?(\d\w\w)$", "$1 $2").ToUpper();
+        }
+        else
+        {
+            postcode = "";
+        }
+
+        return postcode;
+    }
+
+    private void AddPostcodeToPostcodeDb(PostcodeioResponseModel postcodeResponse)

```

Comment:
Can we move private methods below public :)

Reply:
I originally put the private methods below the ones which called them (except for one which I called twice, which I put higher up) - I found it easier for keeping track of the logic while writing the code, but I suppose now it's done it would be easier to have the public methods clearly visible at the top.